File: 200-number-of-islands/200-number-of-islands.cpp
class Solution {
    int m;
    int n;
public:
    // O(mn+e) = O(mn + 4mn) = O(mn) Time, *[O(mn) extra space] and O(mn) stack space
    int numIslands (vector<vector<char>>& grid) {
        m = grid.size(); 
        n = grid[0].size(); 
        // vector<vector<bool>> visited(m, vector<bool>(n, false)); 
        int countIslands = 0; 
        for(int i = 0; i < m; ++i) {
            // [0, m-1] 
            for(int j = 0; j < n; ++j) {
                // [e, n-1] 
                // if(grid[i][j] == '1' && visited[i][j] == false) {
                if(grid[i][j] == '1'){
                    ++countIslands;
                    dfs(grid, i, j);
                }
            }
        }
        return countIslands;
    }
                    
private: 
    void dfs(vector<vector<char>>& grid, int i, int j) {
        grid[i][j] = '0';
        // for each neighbor of vertex (i, j), 
        if(i-1 >= 0 && grid[i-1][j] == '1') {
            dfs(grid, i-1, j); // up cell
        }
        if(j-1 >= 0 && grid[i][j-1] == '1') {
            dfs(grid, i, j-1); // left cell
        }

        if(i+1 < m && grid[i+1][j] == '1') {
            dfs(grid, i+1, j); // down cell
        }
        if(j+1 < n && grid[i][j+1] == '1') {
            dfs (grid, i, j+1); // right cell
        }
        return;
    }
};

File: 200-number-of-islands/200-number-of-islands.py
class Solution:
    def dfs(self, grid, x, y):
        grid[x][y] = '0'
        
        if x >= 1 and grid[x - 1][y] == '1':
            self.dfs(grid, x - 1, y)
            
        if x < len(grid) - 1 and grid[x + 1][y] == '1':
            self.dfs(grid, x + 1, y)
            
        if y >= 1 and grid[x][y - 1] == '1':
            self.dfs(grid, x, y - 1)
            
        if y < len(grid[0]) - 1 and grid[x][y + 1] == '1':
            self.dfs(grid, x, y + 1)
        
    
    def numIslands(self, grid: List[List[str]]) -> int:
        ans = 0
        
        # visited = [[False] * len(grid[0]) for _ in range(len(grid))]
        
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    self.dfs(grid, i, j)
                    ans += 1
        return ans